### ISSUE TITLE
Serializing a vector of classes?

### ISSUE THREAD

**Issue by voidblaze** [nlohmann/json#1136]
I have a vector* of class* (vector<Doctor*> *doctors) and I need to serialize it. The class Doctor has some other vector* of class* that have to be serialized too. The question is, do I need to create to_json and from_json methods for vector*<Doctor*>, and for any other vectors of classes inside Doctor? I have already created the methods for the class doctor and for every other class that needs it, yet I keep getting the "could not find to_json() method in T's namespace". 

I am using MSVS 2017. Thank you in advance for your help!

**Comment by nlohmann** [comment]
Are the methods in the right name space? Can you try without any pointers just to make sure this works?

**Comment by voidblaze** [comment]
It does work when the vector is for the class Doctor and not for a pointer.

**Comment by nlohmann** [comment]
Maybe @theodelrieu has an idea.

**Comment by theodelrieu** [comment]
Pointer to user-defined types conversion is not supported by the library.

But you can do it, you have to specialize `adl_serizalizer` on pointer types (I would recommend supporting `Doctor*` to begin with though).

**Comment by nlohmann** [comment]
Thanks @theodelrieu ! We should put this into the README at some point...

**Comment by theodelrieu** [comment]
In fact it's way simpler, for a single type at least:

```cpp
namespace my_ns {
  void to_json(json& j, const Doctor* doc) {}
}
```

Never thought about that before... So if you want to support both `Doctor*` and `Doctor&`, you have to write two `to_json` overloads.

**Comment by voidblaze** [comment]
But I have more vector to class pointers inside the Doctor class that also need serialization. I have to create a to_json and from_json overload for each of them?

Thank you for the great help btw!

**Comment by theodelrieu** [comment]
>I have to create a to_json and from_json overload for each of them?

Absolutely, that's a bit tedious though. I don't know what is your exact use-case for having vectors of pointers, but I would advise not doing that if you can avoid it.

Unless you're using base classes and inheritance, storing `Doctor` instead of `Doctor*` gives you back a real copy (value semantics, instead of reference semantics).

**Comment by voidblaze** [comment]
I am using inheritance and I am sadly required to have pointers instead of values. I will try to create the methods for class pointers and will report back.

**Comment by voidblaze** [comment]
Not yet, I will have more soon.

**Comment by voidblaze** [comment]
Sadly same error. I created the to_json methods for every class* and I still get that the to_json method doesnt exist.

**Comment by nlohmann** [comment]
Unfortunately, the current error is already an improvement, because it is very difficult to indicate errors there. Can you share some code so we can better understand what's going wrong?

**Comment by voidblaze** [comment]
For sure. I uploaded the peoject here https://github.com/voidblaze/CPP-Appointment-Scheduler

**Comment by theodelrieu** [comment]
I could compile the project with latest GCC on Linux (with `-fpermissive` though, there are some errors unrelated to json).

**Comment by voidblaze** [comment]
Yeah sorry, I forgot to uncomment the line that was causing the error.  Its on data handler, line 191. I believe the errors have something to do with the file opening on line 188, on a path that obviously doesn't exist on your machine. 

I pushed the fixed solution.

**Comment by theodelrieu** [comment]
You're trying to serialize a `std::vector<...>**`, I would advise passing such argument by reference in the first place.

Changing the problematic line to `json jDoctors = *doctors` makes the project compile.

Personally, I would keep pointers for virtual classes, and use references everywhere else.

**Comment by nlohmann** [comment]
Thanks @theodelrieu for helping out!

@voidblaze What would you add to the documentation?
